be/src/storage/compaction_manager.cpp

```diff
@@ -561,8 +560,51 @@ std::unordered_set<CompactionTask*> CompactionManager::get_running_task(const Ta
     return res;
 }
 
+int32_t CompactionManager::compute_max_compaction_concurrency() {
+    int32_t max_task_num = 0;
+    // new compaction framework
+    if (config::base_compaction_num_threads_per_disk >= 0 && config::cumulative_compaction_num_threads_per_disk >= 0) {
+        max_task_num = static_cast<int32_t>(
+                StorageEngine::instance()->get_store_num() *
+                (config::cumulative_compaction_num_threads_per_disk + config::base_compaction_num_threads_per_disk));
+    } else {
+        // When cumulative_compaction_num_threads_per_disk or config::base_compaction_num_threads_per_disk is less than 0,
+        // there is no limit to _max_task_num if max_compaction_concurrency is also less than 0, and here we set maximum value to be 20.
+        max_task_num = std::min(20, static_cast<int32_t>(StorageEngine::instance()->get_store_num() * 5));
+    }
+
+    {
+        std::lock_guard lg(_compact_threads_mutex);
+        if (_max_compaction_concurrency > 0 && _max_compaction_concurrency < max_task_num) {
+            max_task_num = _max_compaction_concurrency;
+        }
+    }
+
+    return max_task_num;
+}
+
+void CompactionManager::set_max_compaction_concurrency(int threads_num) {
+    std::lock_guard lg(_compact_threads_mutex);
+    _max_compaction_concurrency = threads_num;
+}
+
 Status CompactionManager::update_max_threads(int max_threads) {
     if (_compaction_pool != nullptr) {
+        set_max_compaction_concurrency(max_threads);
+        if (max_threads == -1) {
+            _max_task_num = compute_max_compaction_concurrency();
+            return _compaction_pool->update_max_threads(std::max(1, _max_task_num));
+        }
+        if (max_threads == 0) {
+            Status st = _compaction_pool->update_max_threads(0);
+            if (!st.ok()) {
+                return st;
+            }
+            return Status::OK();
```

Comment:
```suggestion
            return _compaction_pool->update_max_threads(0);
```

Haven't done any specific error handling regarding to the return Status value, so just return the status to the caller directly.